File: js/timerEngine.js
import { state, notify } from "./state.js";
import { playTone, playFinishTone } from "./audio.js";
import { vibrate, vibrateFinish } from "./vibration.js";
import { acquireWakeLock, releaseWakeLock } from "./wakeLock.js";

const TICK_MS = 100;
const FINISH_DISPLAY_MS = 4000; // let the full 5-pulse finish alert land before auto-reset

let intervalId = null;

export function canRun() {
  return state.timers.length > 0 && !state.editMode;
}

export function start() {
  if (!canRun()) return;
  if (state.run.status === "idle") {
    state.run.currentIndex = 0;
    state.run.accumulatedMs = 0;
    state.run.loopsRemaining = Math.max(0, state.settings.loopCount - 1);
  }
  state.run.status = "running";
  state.run.segmentStartedAt = Date.now();
  acquireWakeLock();
  startTicking();
  notify();
}

export function stop() {
  if (state.run.status !== "running") return;
  state.run.accumulatedMs += Date.now() - state.run.segmentStartedAt;
  state.run.segmentStartedAt = null;
  state.run.status = "paused";
  stopTicking();
  releaseWakeLock();
  notify();
}

export function reset() {
  state.run.status = "idle";
  state.run.currentIndex = 0;
  state.run.accumulatedMs = 0;
  state.run.segmentStartedAt = null;
  state.run.loopsRemaining = 0;
  stopTicking();
  releaseWakeLock();
  notify();
}

function startTicking() {
  stopTicking();
  intervalId = setInterval(tick, TICK_MS);
}

function stopTicking() {
  if (intervalId) {
    clearInterval(intervalId);
    intervalId = null;
  }
}

function currentElapsedMs() {
  const base = state.run.accumulatedMs;
  if (state.run.status === "running" && state.run.segmentStartedAt != null) {
    return base + (Date.now() - state.run.segmentStartedAt);
  }
  return base;
}

// Progress in [0, 1] for a given timer index, for rendering the fill.
export function getProgress(index) {
  if (state.run.status === "idle") return 0;
  if (index < state.run.currentIndex) return 1;
  if (index > state.run.currentIndex) return 0;
  const timer = state.timers[index];
  if (!timer || timer.durationMs <= 0) return 1;
  return Math.min(1, currentElapsedMs() / timer.durationMs);
}

export function getRemainingMs(index) {
  const timer = state.timers[index];
  if (!timer) return 0;
  if (state.run.status === "idle" || index > state.run.currentIndex) return timer.durationMs;
  if (index < state.run.currentIndex) return 0;
  return Math.max(0, timer.durationMs - currentElapsedMs());
}

function tick() {
  let timer = state.timers[state.run.currentIndex];
  if (!timer) {
    reset();
    return;
  }

  let elapsed = currentElapsedMs();
  const visible = document.visibilityState === "visible";

  while (timer && elapsed >= timer.durationMs) {
    const isLast = state.run.currentIndex >= state.timers.length - 1;
    const isFinalCompletion = isLast && state.run.loopsRemaining <= 0;

    if (visible) {
      if (isFinalCompletion) {
        // End-of-all-loops signal replaces this timer's own alert.
        if (state.settings.finishTone) playFinishTone();
        if (state.settings.finishVibrate) vibrateFinish();
      } else {
        if (timer.tone) playTone(timer.pulseMode);
        if (timer.vibrate) vibrate(timer.pulseMode);
      }
    }

    if (isLast) {
      if (state.run.loopsRemaining > 0) {
        state.run.loopsRemaining -= 1;
        elapsed -= timer.durationMs;
        state.run.currentIndex = 0;
        state.run.accumulatedMs = 0;
        state.run.segmentStartedAt = Date.now() - elapsed;
        timer = state.timers[0];
        continue;
      }

      notify(); // render the final timer as fully complete
      stopTicking();
      releaseWakeLock();
      setTimeout(reset, FINISH_DISPLAY_MS);
      return;
    }

    elapsed -= timer.durationMs;
    state.run.currentIndex += 1;
    state.run.accumulatedMs = 0;
    state.run.segmentStartedAt = Date.now() - elapsed;
    timer = state.timers[state.run.currentIndex];
  }

  notify();
}

File: android/app/src/main/java/com/abregado/joggerloop/engine/TimerEngine.kt
package com.abregado.joggerloop.engine

import com.abregado.joggerloop.data.TimerConfig

/**
 * Pure, framework-free port of the PWA's timerEngine.js. Holds no timers/threads of its
 * own - a caller (the foreground service) drives it by calling [tick] periodically and
 * acting on the returned [TimerEvent]s.
 *
 * [clock] is injected so tests can move time forward instantly instead of sleeping.
 */
class TimerEngine(
    private val timers: List<TimerConfig>,
    private val loopCount: Int,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    var status: RunStatus = RunStatus.IDLE
        private set
    var currentIndex: Int = 0
        private set
    var loopsRemaining: Int = 0
        private set

    private var accumulatedMs: Long = 0L
    private var segmentStartedAt: Long? = null

    fun start() {
        if (timers.isEmpty()) return
        if (status == RunStatus.IDLE || status == RunStatus.FINISHED) {
            currentIndex = 0
            accumulatedMs = 0
            loopsRemaining = (loopCount - 1).coerceAtLeast(0)
        }
        status = RunStatus.RUNNING
        segmentStartedAt = clock()
    }

    fun stop() {
        if (status != RunStatus.RUNNING) return
        accumulatedMs += clock() - (segmentStartedAt ?: clock())
        segmentStartedAt = null
        status = RunStatus.PAUSED
    }

    fun reset() {
        status = RunStatus.IDLE
        currentIndex = 0
        accumulatedMs = 0
        segmentStartedAt = null
        loopsRemaining = 0
    }

    /** Progress in [0, 1] for the given timer index, for rendering a fill. */
    fun getProgress(index: Int): Float {
        if (status == RunStatus.IDLE) return 0f
        if (status == RunStatus.FINISHED) return 1f
        if (index < currentIndex) return 1f
        if (index > currentIndex) return 0f
        val timer = timers.getOrNull(index) ?: return 1f
        if (timer.durationMs <= 0) return 1f
        return (currentElapsedMs().toFloat() / timer.durationMs).coerceAtMost(1f)
    }

    fun getRemainingMs(index: Int): Long {
        val timer = timers.getOrNull(index) ?: return 0L
        if (status == RunStatus.FINISHED) return 0L
        if (status == RunStatus.IDLE || index > currentIndex) return timer.durationMs
        if (index < currentIndex) return 0L
        return (timer.durationMs - currentElapsedMs()).coerceAtLeast(0L)
    }

    private fun currentElapsedMs(): Long {
        val base = accumulatedMs
        val startedAt = segmentStartedAt
        return if (status == RunStatus.RUNNING && startedAt != null) {
            base + (clock() - startedAt)
        } else {
            base
        }
    }

    /**
     * Advances the engine to "now" and returns whatever [TimerEvent]s occurred along the
     * way. A single call can return multiple events if it's been long enough since the
     * last tick to cross more than one timer/loop boundary (e.g. after the process was
     * briefly suspended) - every crossing still fires its alert, since (unlike the PWA)
     * this engine has no reason to suppress alerts for being "backgrounded."
     *
     * No-op once [status] is [RunStatus.FINISHED]: the finish alert already fired for
     * this run, and repeated calls during the frozen display window must not replay it.
     */
    fun tick(): List<TimerEvent> {
        if (status != RunStatus.RUNNING) return emptyList()

        if (currentIndex !in timers.indices) {
            reset()
            return emptyList()
        }

        var timer: TimerConfig = timers[currentIndex]
        var elapsed = currentElapsedMs()
        val events = mutableListOf<TimerEvent>()

        while (elapsed >= timer.durationMs) {
            val isLast = currentIndex >= timers.lastIndex
            val isFinalCompletion = isLast && loopsRemaining <= 0

            events += if (isFinalCompletion) TimerEvent.AllLoopsFinished else TimerEvent.TimerAlert(timer)

            if (isLast) {
                if (loopsRemaining > 0) {
                    loopsRemaining -= 1
                    elapsed -= timer.durationMs
                    currentIndex = 0
                    accumulatedMs = 0
                    segmentStartedAt = clock() - elapsed
                    timer = timers[0]
                    continue
                }

                status = RunStatus.FINISHED
                return events
            }

            elapsed -= timer.durationMs
            currentIndex += 1
            accumulatedMs = 0
            segmentStartedAt = clock() - elapsed
            timer = timers[currentIndex]
        }

        return events
    }
}
